src/pg_query_summary.h

```diff
@@ -0,0 +1,90 @@
+#ifndef PG_QUERY_SUMMARY_H
+#define PG_QUERY_SUMMARY_H
+
+#include "pg_query.h"
+#include "pg_query_internal.h"
+
+#include "nodes/primnodes.h"
+
+#include <unistd.h>
+
+typedef enum {
+	CONTEXT_NONE = 0,
+	CONTEXT_SELECT = 1,
+	CONTEXT_DML = 2,
+	CONTEXT_DDL = 3,
+	CONTEXT_CALL = 4,
+} ContextType;
+
+typedef struct {
+	char *name;
+	char *schema_name;
+	char *table_name;
+	ContextType context;
+} SummaryTable;
+
+typedef struct {
+	char *key;
+	char *value;
+} SummaryAlias;
+
+typedef struct {
+	char *name;
+	char *function_name;
+	char *schema_name; // optional
+	ContextType context;
+} SummaryFunction;
+
+typedef struct {
+	char *schema_name;
+	char *table_name;
+	char *column;
+} FilterColumn;
+
+typedef struct {
+	List *tables; // List of SummaryTable*
+	List *aliases; // List of SummaryAlias*
+	List *cte_names;
+	List *functions; // List of SummaryFunction*
+	List *filter_columns; // List of FilterColumn*
+	List *statement_types; // List of char*
+
+	// UNIMPLEMENTED(truncated_query): char *truncated_query;
+} Summary;
+
+typedef struct {
+	List* tables; // List of Table*
+	List *aliases; // List of SummaryAlias*
+	List* functions; // List of Function*
+	List* subselect_items;
+	List* range_vars;
+	List* cte_names; // list of CommonTableExpr
+	List* filter_columns; // List of FilterColumn*
+	bool save_filter_columns; // If true, we should be adding to filter columns.
+} WalkState;
+
+typedef struct {
+	RangeVar* node;
+	ContextType context;
+} RangeVarWithContext;
+
+// This is like PgQuerySummaryParseResult, but the summary is not wrapped
+// up in protobuf.
+typedef struct {
+	Summary summary;
+	char* stderr_buffer;
+	PgQueryError* error;
+	MemoryContext ctx;
+} PgQuerySummaryParseResultDirect;
+
+PgQueryError* pg_query_summary_statement_walk(Summary *summary, Node *node);
+
+// This is like pg_query_summary(), but returns a (non-protobuf)
+// PgQuerySummaryParseResultDirect instead of the usual
+// PgQuerySummaryParseResult which converts the summary to protobuf.
+//
+// This is used for tests, to avoid having to process protobuf in C.
+PgQuerySummaryParseResultDirect pg_query_summary_direct(const char* input, int parser_options, int truncate_limit);
```

Comment:
Maybe `_internal` and `PgQuerySummaryParseResultInternal` would be better, to make it clear to the caller we don't expect them to use it.